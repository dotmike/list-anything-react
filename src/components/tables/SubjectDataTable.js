import React from "react";
import { Table, Loader } from "semantic-ui-react";
import PropTypes from "prop-types";
import sortBy from "lodash.sortby";

import renderFieldComponent from "../../utils/renderFieldComponent";

class SubjectDataTable extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      column: null,
      data: props.subjectDataArray,
      direction: null
    };
  }

  componentWillReceiveProps = nextProps => {
    if (this.props.subjectDataArray !== nextProps.subjectDataArray) {
      this.setState({ data: nextProps.subjectDataArray });
    }
  };

  // Talvez seja necessário lidar com outros tipos também...
  getCellWidth = fieldType =>
    fieldType === "url_input_img" ? { width: 2 } : {};

  handleSort = clickedColumn => () => {
    const { column, data, direction } = this.state;

    if (column !== clickedColumn) {
      this.setState({
        column: clickedColumn,
        data: sortBy(data, [o => o.data[clickedColumn].value]),
        direction: "ascending"
      });
    } else {
      this.setState({
        data: data.reverse(),
        direction: direction === "ascending" ? "descending" : "ascending"
      });
    }
  };

  renderCell = (subjectData, field) => {
    const data = subjectData.data[field._id];
    return (
      <Table.Cell
        {...this.getCellWidth(field.field_type)}
        collapsing
        key={data._id}
      >
        {this.renderField(data, field)}
      </Table.Cell>
    );
  };

  renderField = (data, field) => {
    const fieldData = {
      key: data._id,
      value: data.value,
      showFieldDescription: false,
      field
    };
    return renderFieldComponent(fieldData);
  };

  render() {
    const { loading, onTableRowClick, fields } = this.props;
    const { column, direction, data } = this.state;

    return (
      <Table celled compact="very" selectable sortable>
        <Table.Header>
          <Table.Row>
            {fields.map(
              field =>
                field.show_in_list && (
                  <Table.HeaderCell
                    key={field._id}
                    sorted={column === field._id ? direction : null}
                    onClick={this.handleSort(field._id)}
                  >
                    {field.description}
                  </Table.HeaderCell>
                )
            )}
          </Table.Row>
        </Table.Header>
        <Table.Body style={{ cursor: "pointer" }}>
          {loading && (
            <Table.Row>
              <Table.Cell width={16}>
                <Loader active inline="centered" />
              </Table.Cell>
            </Table.Row>
          )}
          {!loading &&
            data.map(sd => (
              <Table.Row
                key={sd._id}
                to={`/subject-data/${sd._id}`}
                onClick={onTableRowClick}
              >
                {fields.map(
                  field => field.show_in_list && this.renderCell(sd, field)
                )}
              </Table.Row>
            ))}
        </Table.Body>
      </Table>
    );
  }
}

SubjectDataTable.propTypes = {
  // ownProps
  loading: PropTypes.bool.isRequired,
  fields: PropTypes.arrayOf(
    PropTypes.shape({
      description: PropTypes.string.isRequired
    })
  ).isRequired,
  subjectDataArray: PropTypes.arrayOf(
    PropTypes.shape({
      _id: PropTypes.string.isRequired
    })
  ).isRequired,
  onTableRowClick: PropTypes.func.isRequired
};

SubjectDataTable.defaultProps = {
  fieldHash: {}
};

export default SubjectDataTable;
