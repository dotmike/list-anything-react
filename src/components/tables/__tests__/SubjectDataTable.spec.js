import SubjectDataTable from "../SubjectDataTable";

const testData = {
  fields: [
    {
      _id: "0",
      description: "Test Description",
      field_type: "test_field",
      show_in_list: true
    }
  ],
  subjectDataArray: [
    {
      _id: "1",
      tabId: "2",
      data: {
        "0": {
          _id: "3",
          fieldId: "0",
          value: "Test value"
        }
      }
    }
  ]
};

const setup = (propOverrides = {}) => {
  const props = {
    loading: false,
    fields: testData.fields,
    subjectDataArray: testData.subjectDataArray,
    onTableRowClick: jest.fn(),
    ...propOverrides
  };

  return {
    props,
    // Lembrar de sempre chamar `unmount` quando usar esse wrapper \/
    wrapperMount: wrapperMount(SubjectDataTable, props)
  };
};

describe("SubjectDataTable", () => {
  it("renders correctly when `loading` is true", () => {
    const { wrapperMount: wrapper } = setup({
      loading: true
    });
    expect(wrapper()).toMatchSnapshot();

    wrapper().unmount();
  });

  describe("when `loading` is false", () => {
    it("renders correctly when `subjectDataArray` is empty", () => {
      const { wrapperMount: wrapper } = setup({
        subjectDataArray: []
      });
      expect(wrapper()).toMatchSnapshot();

      wrapper().unmount();
    });

    it("renders correctly when `subjectDataArray` is not empty", () => {
      const { wrapperMount: wrapper } = setup();
      expect(wrapper()).toMatchSnapshot();

      wrapper().unmount();
    });

    it("calls `onTableRowClick` when clicking on a `TableBody > TableRow`", () => {
      const { wrapperMount: wrapper, props } = setup();
      expect(wrapper()).toMatchSnapshot();

      wrapper()
        .find("TableBody")
        .find("TableRow")
        .first()
        .simulate("click");

      expect(props.onTableRowClick).toHaveBeenCalled();

      wrapper().unmount();
    });
  });
});
