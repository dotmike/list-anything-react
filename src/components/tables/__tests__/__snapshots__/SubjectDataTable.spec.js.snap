// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectDataTable renders correctly when \`loading\` is true 1`] = `
<SubjectDataTable
  fields={
    Array [
      Object {
        "_id": "0",
        "description": "Test Description",
        "field_type": "test_field",
        "show_in_list": true,
      },
    ]
  }
  loading={true}
  onTableRowClick={[Function]}
  subjectDataArray={
    Array [
      Object {
        "_id": "1",
        "data": Object {
          "0": Object {
            "_id": "3",
            "fieldId": "0",
            "value": "Test value",
          },
        },
        "tabId": "2",
      },
    ]
  }
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={
          Array [
            Object {
              "_id": "1",
              "data": Object {
                "0": Object {
                  "_id": "3",
                  "fieldId": "0",
                  "value": "Test value",
                },
              },
              "tabId": "2",
            },
          ]
        }
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="0"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Test Description
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableCell
                  as="td"
                  width={16}
                >
                  <td
                    className="sixteen wide"
                  >
                    <Loader
                      active={true}
                      inline="centered"
                    >
                      <div
                        className="ui active centered inline loader"
                      />
                    </Loader>
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SubjectDataTable>
`;

exports[`SubjectDataTable when \`loading\` is false calls \`onTableRowClick\` when clicking on a \`TableBody > TableRow\` 1`] = `
<SubjectDataTable
  fields={
    Array [
      Object {
        "_id": "0",
        "description": "Test Description",
        "field_type": "test_field",
        "show_in_list": true,
      },
    ]
  }
  loading={false}
  onTableRowClick={[Function]}
  subjectDataArray={
    Array [
      Object {
        "_id": "1",
        "data": Object {
          "0": Object {
            "_id": "3",
            "fieldId": "0",
            "value": "Test value",
          },
        },
        "tabId": "2",
      },
    ]
  }
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={
          Array [
            Object {
              "_id": "1",
              "data": Object {
                "0": Object {
                  "_id": "3",
                  "fieldId": "0",
                  "value": "Test value",
                },
              },
              "tabId": "2",
            },
          ]
        }
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="0"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Test Description
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
              key="1"
              onClick={[Function]}
              to="/subject-data/1"
            >
              <tr
                className=""
                onClick={[Function]}
                to="/subject-data/1"
              >
                <TableCell
                  as="td"
                  collapsing={true}
                  key="3"
                >
                  <td
                    className="collapsing"
                  >
                    <div
                      key="0"
                    />
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SubjectDataTable>
`;

exports[`SubjectDataTable when \`loading\` is false renders correctly when \`subjectDataArray\` is empty 1`] = `
<SubjectDataTable
  fields={
    Array [
      Object {
        "_id": "0",
        "description": "Test Description",
        "field_type": "test_field",
        "show_in_list": true,
      },
    ]
  }
  loading={false}
  onTableRowClick={[Function]}
  subjectDataArray={Array []}
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={Array []}
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="0"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Test Description
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableCell
                  as="td"
                  width={16}
                >
                  <td
                    className="sixteen wide"
                  >
                    Nothing was found.
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SubjectDataTable>
`;

exports[`SubjectDataTable when \`loading\` is false renders correctly when \`subjectDataArray\` is not empty 1`] = `
<SubjectDataTable
  fields={
    Array [
      Object {
        "_id": "0",
        "description": "Test Description",
        "field_type": "test_field",
        "show_in_list": true,
      },
    ]
  }
  loading={false}
  onTableRowClick={[Function]}
  subjectDataArray={
    Array [
      Object {
        "_id": "1",
        "data": Object {
          "0": Object {
            "_id": "3",
            "fieldId": "0",
            "value": "Test value",
          },
        },
        "tabId": "2",
      },
    ]
  }
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={
          Array [
            Object {
              "_id": "1",
              "data": Object {
                "0": Object {
                  "_id": "3",
                  "fieldId": "0",
                  "value": "Test value",
                },
              },
              "tabId": "2",
            },
          ]
        }
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="0"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Test Description
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
              key="1"
              onClick={[Function]}
              to="/subject-data/1"
            >
              <tr
                className=""
                onClick={[Function]}
                to="/subject-data/1"
              >
                <TableCell
                  as="td"
                  collapsing={true}
                  key="3"
                >
                  <td
                    className="collapsing"
                  >
                    <div
                      key="0"
                    />
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SubjectDataTable>
`;
